assets/js/wpjm-stats.js

```diff
@@ -4,7 +4,148 @@ import domReady from '@wordpress/dom-ready';
+	function debounce(func, delay) {
+		let timeoutId;
+
+		return function (...args) {
+			clearTimeout(timeoutId);
+
+			timeoutId = setTimeout(() => {
+				func(...args);
+			}, delay);
+		};
+	}
+
+	function createStatsQueue() {
+		const alreadySent = {};
+		let queue = [];
+
+		const logThem = debounce( function ( listingIds ) {
+			const stats = listingIds.map( function ( id ) {
+				alreadySent[id] = true;
+				return { name: 'job_listing_impressions', post_id: id };
+			});
+			console.log( 'sending stats:', stats );
+			// queue = [];
+			return wpjmLogStats( stats ).finally( function () {
+				queue = [];
+			});
+		}, 500 );
+
+		return {
+			queueListingImpressionStats: function ( listingIds ) {
+				listingIds.forEach( function (listingId ) {
+					if ( ! alreadySent[listingId] ) {
+						queue.push(listingId);
+					}
+				} );
+				logThem( queue );
+
+			}
+		};
+	}
+
+	function observeForVisibility( jobListingContainer, jobListingElement, visibleCallback, alreadyViewedListings ) {
+		const options = {
+			root: null,
+			rootMargin: "0px",
+			threshold: 1.0,
+		};
+
+		const observer = new IntersectionObserver(function ( entries ) {
+			entries.forEach(function ( entry ) {
+				if ( entry.isIntersecting && entry.intersectionRatio > 0.99 ) {
+					const node = entry.target;
+					if ( 1 === node.nodeType && node.classList.contains( 'job_listing' ) ) {
+						const nodeId = node.dataset.id;
+						if ( ! alreadyViewedListings[nodeId] ) {
+							alreadyViewedListings[nodeId] = true;
+							visibleCallback( node );
+						}
+					}
+					node.classList.add( 'viewed' );
+					observer.unobserve( node );
+				}
+			} );
+		}, options );
+
+		observer.observe( jobListingElement );
+	}
+
+	function waitForNextVisibleListing( listingVisibleCallback ) {
+		const jobListingsContainer = document.querySelector( 'ul.job_listings' );
+		const config = { childList: true };
+		const alreadyViewed = {};
+
+		const observer = new MutationObserver(function ( mutations ) {
+			mutations.forEach(function ( mutation ) {
+				mutation.addedNodes.forEach( function ( node ) {
+					if ( 1 === node.nodeType && node.classList.contains( 'job_listing' ) && ! node.classList.contains( 'viewed' ) ) {
+						observeForVisibility( jobListingsContainer, node, listingVisibleCallback, alreadyViewed );
+					}
+				} );
+			} );
+		} );
+
+		observer.observe( jobListingsContainer, config );
+	}
+
+	function waitForSelector( selector ) {
+		return new Promise(function ( resolve ) {
+			if (document.querySelector(selector)) {
+				return resolve(document.querySelector(selector));
+			}
+
+			const observer = new MutationObserver(function () {
+				if (document.querySelector(selector)) {
+					observer.disconnect();
+					resolve(document.querySelector(selector));
+				}
+			});
+
+			observer.observe(document.documentElement, {
+				childList: true,
+				subtree: true
+			});
+		});
+	}
+
+	function existsAndCallable( globalAsString ) {
```

Comment:
Instead of defining the callbacks as `'WPJMStats.initListingImpression',` and needing this kind of check, I'd keep it fixed to one object, define it in PHP as `'initListingImpression'`, and call it like `WPJMStats[ stat.js_callback ]` in JS. 

Together with the DOM-event based stats in #2763, maybe we could generalize these different setup needs as the stat `type`. Something like:

```js
WPJMStats.types = {
  impression: ( stat ) => { /* set up observer */ }, 
  domEvent: ( stat ) => { /* addEventListener setups */ }
}

// Stat is set up like this:
WPJMStats.types[ stat.type ]?( stat )
```

Reply:
The reason for making this so generic is I thought it woudl be a cool idea to allow for arbitrary namespaces/objects (catering for 3rd-party stuff)

Reply:
Yeah, in the example above a plugin could add its thing like `WPJMStats.types.initAlerts = ...`